hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -0,0 +1,317 @@
+  */
+case class HoodieFileIndex(
+     spark: SparkSession,
+     basePath: String,
+     schemaSpec: Option[StructType],
+     options: Map[String, String])
+  extends FileIndex with Logging {
+
+  @transient private val hadoopConf = spark.sessionState.newHadoopConf()
+  private lazy val metaClient = HoodieTableMetaClient
+    .builder().setConf(hadoopConf).setBasePath(basePath).build()
+
+  @transient private val queryPath = new Path(options.getOrElse("path", "'path' option required"))
+  /**
+    * Get the schema of the table.
+    */
+  lazy val schema: StructType = schemaSpec.getOrElse({
+    val schemaUtil = new TableSchemaResolver(metaClient)
+    SchemaConverters.toSqlType(schemaUtil.getTableAvroSchema)
+      .dataType.asInstanceOf[StructType]
+  })
+
+  /**
+    * Get the partition schema from the hoodie.properties.
+    */
+  private lazy val _partitionSchemaFromProperties: StructType = {
+    val tableConfig = metaClient.getTableConfig
+    val partitionColumns = tableConfig.getPartitionColumns
+    val nameFieldMap = schema.fields.map(filed => filed.name -> filed).toMap
+
+    if (partitionColumns.isPresent) {
+      val partitionFields = partitionColumns.get().map(column =>
+        nameFieldMap.getOrElse(column, throw new IllegalArgumentException(s"Cannot find column " +
+          s"$column in the schema[${schema.fields.mkString(",")}]")))
+      new StructType(partitionFields)
+    } else { // If the partition columns have not stored in hoodie.properites(the table that was
+      // created earlier), we trait it as a none-partitioned table.
+      new StructType()
+    }
+  }
+
+  @transient @volatile private var fileSystemView: HoodieTableFileSystemView = _
+  @transient @volatile private var cachedAllInputFiles: Array[HoodieBaseFile] = _
+  @transient @volatile private var cachedFileSize: Long = 0L
+  @transient @volatile private var cachedAllPartitionPaths: Seq[PartitionPath] = _
+
+  @volatile private var queryAsNonePartitionedTable: Boolean = _
+
+  refresh()
+
+  override def rootPaths: Seq[Path] = queryPath :: Nil
+
+  override def listFiles(partitionFilters: Seq[Expression],
+                         dataFilters: Seq[Expression]): Seq[PartitionDirectory] = {
+    if (queryAsNonePartitionedTable) { // Read as None Partitioned table.
+      Seq(PartitionDirectory(InternalRow.empty, allFiles))
+    } else {
+      // Prune the partition path by the partition filters
+      val prunedPartitions = prunePartition(cachedAllPartitionPaths, partitionFilters)
+      prunedPartitions.map { partition =>
+        val fileStatues = fileSystemView.getLatestBaseFiles(partition.partitionPath).iterator()
+          .asScala.toSeq
+          .map(_.getFileStatus)
+        PartitionDirectory(partition.values, fileStatues)
+      }
+    }
+  }
+
+  override def inputFiles: Array[String] = {
+    cachedAllInputFiles.map(_.getFileStatus.getPath.toString)
+  }
+
+  override def refresh(): Unit = {
```

Comment:
- Do you know when this `refresh()` is called apart from the internal call you make from within this index class at the time of index creation ?
- The `InMemoryFileIndex` mantains `refresh()` and `refresh0()`. The `refresh0()` is called from within the index class at the time of creation. While `refresh()` is called from other places too. Do you know why we don't need to follow similar model ?

I will do more reading of the spark code to understand this part as well, but if you already know the answers you can respond to the questions I have.

Reply:
Hi @umehrot2 , AFAIK, The `refresh()` will be called the follow case:
1、The Refresh Table Command.
2、After write finish, e.g `InsertIntoHadoopFsRelationCommand` will call `fileIndex.foreach(_.refresh())`.

> The InMemoryFileIndex mantains refresh() and refresh0(). The refresh0() is called from within the index class at the time of creation. While refresh() is called from other places too. Do you know why we don't need to follow similar model ?

The `InMemoryFileIndex` has some logical about the `fileStatusCache` in the `refresh`. So it has to distinguish between `refresh` and `refresh0`. But in the HoodieFileIndex, there is no such case.